{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#! /usr/bin/env python\n",
    "# -*- coding: utf-8 -*-\n",
    "\n",
    "'''\n",
    "This module contains classes for simulation and inference for a binary branching process with mutation\n",
    "in which the tree is collapsed to nodes that count the number of clonal leaves of each type\n",
    "'''\n",
    "\n",
    "from __future__ import division, print_function\n",
    "import scipy, warnings, random\n",
    "try:\n",
    "    import cPickle as pickle\n",
    "except:\n",
    "    import pickle\n",
    "from scipy.misc import logsumexp\n",
    "from scipy.optimize import minimize, check_grad\n",
    "\n",
    "import matplotlib\n",
    "matplotlib.use('PDF')\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib import rc, ticker\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from scipy.stats import probplot\n",
    "from ete3 import TreeNode, NodeStyle, TreeStyle, TextFace, add_face_to_node, CircleFace, faces, AttrFace\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna\n",
    "\n",
    "#scipy.seterr(all='raise')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class LeavesAndClades():\n",
    "    '''\n",
    "    This is a base class for simulating, and computing likelihood for, an infinite type branching\n",
    "    process with branching probability p, mutation probability q, and we collapse mutant clades off the\n",
    "    root type and consider just the number of clone leaves, c, and mutant clades, m.\n",
    "\n",
    "      /\\\n",
    "     /\\ ^          (3)\n",
    "      /\\     ==>   / \\\\\n",
    "       /\\\\\n",
    "        ^\n",
    "    '''\n",
    "    def __init__(self, params=None, c=None, m=None):\n",
    "        '''initialize with branching probability p and mutation probability q, both in the unit interval'''\n",
    "        if params is not None:\n",
    "            p, q = params\n",
    "            if not (0 <= p <= 1 and 0 <= q <= 1):\n",
    "                raise ValueError('p and q must be in the unit interval')\n",
    "        self._nparams = 2#len(params)\n",
    "        self.params = params\n",
    "        if c is not None or m is not None:\n",
    "            if not (c >= 0) and (m >= 0) and (c+m > 0):\n",
    "                raise ValueError('c and m must be nonnegative integers summing greater than zero')\n",
    "            self.c = c\n",
    "            self.m = m\n",
    "\n",
    "    def simulate(self):\n",
    "        '''simulate the number of clone leaves and mutant clades off a root node'''\n",
    "        if self.params[0]>=.5:\n",
    "            warnings.warn('p >= .5 is not subcritical, tree simulations not garanteed to terminate')\n",
    "        if self.params is None:\n",
    "            raise ValueError('params must be defined for simulation\\n')\n",
    "\n",
    "        # let's track the tree in breadth first order, listing number clone and mutant descendants of each node\n",
    "        # mutant clades terminate in this view\n",
    "        cumsum_clones = 0\n",
    "        len_tree = 0\n",
    "        self.c = 0\n",
    "        self.m = 0\n",
    "        # while termination condition not met\n",
    "        while cumsum_clones > len_tree - 1:\n",
    "            if random.random() < self.params[0]:\n",
    "                mutants = sum(random.random() < self.params[1] for child in range(2))\n",
    "                clones = 2 - mutants\n",
    "                self.m += mutants\n",
    "            else:\n",
    "                mutants = 0\n",
    "                clones = 0\n",
    "                self.c += 1\n",
    "            cumsum_clones += clones\n",
    "            len_tree += 1\n",
    "        assert cumsum_clones == len_tree - 1\n",
    "\n",
    "    f_hash = {} # <--- class variable for hashing calls to the following function\n",
    "    def f(self, params):\n",
    "        '''\n",
    "        Probability of getting c leaves that are clones of the root and m mutant clades off\n",
    "        the root line, given branching probability p and mutation probability q\n",
    "        Also returns gradient wrt (p, q)\n",
    "        Computed by dynamic programming\n",
    "        '''\n",
    "        p, q = params\n",
    "        c, m = self.c, self.m\n",
    "        if (p, q, c, m) not in LeavesAndClades.f_hash:\n",
    "            if c==m==0 or (c==0 and m==1):\n",
    "                f_result = 0\n",
    "                dfdp_result = 0\n",
    "                dfdq_result = 0\n",
    "            elif c==1 and m==0:\n",
    "                f_result = 1-p\n",
    "                dfdp_result = -1\n",
    "                dfdq_result = 0\n",
    "            elif c==0 and m==2:\n",
    "                f_result = p*q**2\n",
    "                dfdp_result = q**2\n",
    "                dfdq_result = 2*p*q\n",
    "            else:\n",
    "                if m >= 1:\n",
    "                    neighbor = LeavesAndClades(params=params, c=c, m=m-1)\n",
    "                    neighbor_f, (neighbor_dfdp, neighbor_dfdq) = neighbor.f(params)\n",
    "                    f_result = 2*p*q*(1-q)*neighbor_f\n",
    "                    dfdp_result =   2*q*(1-q) * neighbor_f + \\\n",
    "                                  2*p*q*(1-q) * neighbor_dfdp\n",
    "                    dfdq_result = (2*p - 4*p*q) * neighbor_f + \\\n",
    "                                   2*p*q*(1-q)  * neighbor_dfdq\n",
    "                else:\n",
    "                    f_result = 0.\n",
    "                    dfdp_result = 0.\n",
    "                    dfdq_result = 0.\n",
    "                for cx in range(c+1):\n",
    "                    for mx in range(m+1):\n",
    "                        if (not (cx==0 and mx==0)) and (not (cx==c and mx==m)):\n",
    "                            neighbor1 = LeavesAndClades(params=params, c=cx, m=mx)\n",
    "                            neighbor2 = LeavesAndClades(params=params, c=c-cx, m=m-mx)\n",
    "                            neighbor1_f, (neighbor1_dfdp, neighbor1_dfdq) = neighbor1.f(params)\n",
    "                            neighbor2_f, (neighbor2_dfdp, neighbor2_dfdq) = neighbor2.f(params)\n",
    "                            f_result += p*(1-q)**2*neighbor1_f*neighbor2_f\n",
    "                            dfdp_result +=   (1-q)**2 * neighbor1_f    * neighbor2_f + \\\n",
    "                                           p*(1-q)**2 * neighbor1_dfdp * neighbor2_f + \\\n",
    "                                           p*(1-q)**2 * neighbor1_f    * neighbor2_dfdp\n",
    "                            dfdq_result += -2*p*(1-q) * neighbor1_f    * neighbor2_f + \\\n",
    "                                           p*(1-q)**2 * neighbor1_dfdq * neighbor2_f + \\\n",
    "                                           p*(1-q)**2 * neighbor1_f    * neighbor2_dfdq\n",
    "            LeavesAndClades.f_hash[(p, q, c, m)] = (f_result, scipy.array([dfdp_result, dfdq_result]))\n",
    "        return LeavesAndClades.f_hash[(p, q, c, m)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class CollapsedTree(LeavesAndClades):\n",
    "    '''\n",
    "    Here's a derived class for a collapsed tree, where we recurse into the mutant clades\n",
    "          (4)\n",
    "         / | \\\\\n",
    "       (3)(1)(2)\n",
    "           |   \\\\\n",
    "          (2)  (1)\n",
    "    '''\n",
    "    def __init__(self, params=None, tree=None, frame=None):\n",
    "        '''\n",
    "        For intialization, either params or tree (or both) must be provided\n",
    "        params: offspring distribution parameters\n",
    "        tree: ete tree with frequency node feature. If uncollapsed, it will be collapsed\n",
    "        frame: tranlation frame, with default None, no tranlation attempted\n",
    "        '''\n",
    "        #if params is None and tree is None:\n",
    "        #    raise ValueError('either params or tree (or both) must be provided')\n",
    "        LeavesAndClades.__init__(self, params=params)\n",
    "        if frame is not None and frame not in (1, 2, 3):\n",
    "            raise RuntimeError('frame must be 1, 2, 3, or None')\n",
    "        self.frame = frame\n",
    "        if tree is not None:\n",
    "            self.tree = tree.copy()\n",
    "            if 0 in (node.dist for node in tree.iter_descendants()):\n",
    "                # iterate over the tree below root and collapse edges of zero length\n",
    "                for node in self.tree.get_descendants():\n",
    "                    if node.dist == 0:\n",
    "                        node.up.frequency += node.frequency\n",
    "                        node.delete(prevent_nondicotomic=False)\n",
    "            assert sum(node.frequency for node in tree.traverse()) == sum(node.frequency for node in self.tree.traverse())\n",
    "            if 'sequence' in tree.features and len(set([node.sequence for node in self.tree.traverse()])) != sum(1 for _ in self.tree.traverse()):\n",
    "                raise RuntimeError('repeated sequences in collapsed tree, possible backmutation')\n",
    "        else:\n",
    "            self.tree = tree\n",
    "\n",
    "\n",
    "    def l(self, params, sign=1):\n",
    "        '''\n",
    "        log likelihood of params, conditioned on collapsed tree, and its gradient wrt params\n",
    "        optional parameter sign must be 1 or -1, with the latter useful for MLE by minimization\n",
    "        '''\n",
    "        if self.tree is None:\n",
    "            raise ValueError('tree data must be defined to compute likelihood')\n",
    "        if sign not in (-1, 1):\n",
    "            raise ValueError('sign must be 1 or -1')\n",
    "        leaves_and_clades_list = [LeavesAndClades(c=node.frequency, m=len(node.children)) for node in self.tree.traverse()]\n",
    "        if leaves_and_clades_list[0].c == 0 and leaves_and_clades_list[0].m == 1 and leaves_and_clades_list[0].f(params)[0] == 0:\n",
    "            # if unifurcation not possible under current model, omit root\n",
    "            leaves_and_clades_list = leaves_and_clades_list[1:]\n",
    "        # extract vector of function values and gradient components\n",
    "        f_data = [leaves_and_clades.f(params) for leaves_and_clades in leaves_and_clades_list]\n",
    "        fs = scipy.array([[x[0]] for x in f_data])\n",
    "        logf = scipy.log(fs).sum()\n",
    "        grad_fs = scipy.array([x[1] for x in f_data])\n",
    "        grad_logf = (grad_fs/fs).sum(axis=0)\n",
    "        return sign*logf, sign*grad_logf\n",
    "\n",
    "    def mle(self, **kwargs):\n",
    "        '''\n",
    "        Maximum likelihood estimate for params given tree\n",
    "        updates params if not None\n",
    "        returns optimization result\n",
    "        '''\n",
    "        # random initalization\n",
    "        x_0 = (random.random(), random.random())\n",
    "        #x_0 = (.5, .5)\n",
    "        bounds = ((.01, .99), (.001, .999))\n",
    "        kwargs['sign'] = -1\n",
    "        grad_check = check_grad(lambda x: self.l(x, **kwargs)[0], lambda x: self.l(x, **kwargs)[1], (.4, .5))\n",
    "        if grad_check > 1e-3:\n",
    "            warnings.warn('gradient mismatches finite difference approximation by {}'.format(grad_check), RuntimeWarning)\n",
    "        result = minimize(lambda x: self.l(x, **kwargs), x0=x_0, jac=True, method='L-BFGS-B', options={'ftol':1e-10}, bounds=bounds)\n",
    "        # update params if None and optimization successful\n",
    "        if not result.success:\n",
    "            warnings.warn('optimization not sucessful, '+result.message, RuntimeWarning)\n",
    "        elif self.params is None:\n",
    "            self.params = result.x.tolist()\n",
    "        return result\n",
    "\n",
    "    def simulate(self):\n",
    "        '''\n",
    "        simulate a collapsed tree given params\n",
    "        replaces existing tree data member with simulation result, and returns self\n",
    "        '''\n",
    "        if self.params is None:\n",
    "            raise ValueError('params must be defined for simulation')\n",
    "\n",
    "        # initiate by running a LeavesAndClades simulation to get the number of clones and mutants\n",
    "        # in the root node of the collapsed tree\n",
    "        LeavesAndClades.simulate(self)\n",
    "        self.tree = TreeNode()\n",
    "        self.tree.add_feature('frequency', self.c)\n",
    "        if self.m == 0:\n",
    "            return self\n",
    "        for _ in range(self.m):\n",
    "            # ooooh, recursion\n",
    "            child = CollapsedTree(params=self.params, frame=self.frame).simulate().tree\n",
    "            child.dist = 1\n",
    "            self.tree.add_child(child)\n",
    "\n",
    "        return self\n",
    "\n",
    "\n",
    "    def __str__(self):\n",
    "        '''return a string representation for printing'''\n",
    "        return 'params = ' + str(self.params)+ '\\ntree:\\n' + str(self.tree)\n",
    "\n",
    "    def render(self, outfile):\n",
    "        '''render to image file, filetype inferred from suffix, svg for color images'''\n",
    "        for node in self.tree.traverse():\n",
    "            nstyle = NodeStyle()\n",
    "            if node.frequency == 0:\n",
    "                nstyle['size'] = 5\n",
    "                nstyle['fgcolor'] = 'grey'\n",
    "            else:\n",
    "                nstyle['size'] = 3*2*scipy.sqrt(scipy.pi*node.frequency)\n",
    "                nstyle['fgcolor'] = 'black'\n",
    "            if node.up is not None:\n",
    "                if set(node.sequence.upper()) == set('ACGT'):\n",
    "                    if self.frame is not None:\n",
    "                        aa = Seq(node.sequence[(self.frame-1):(self.frame-1+(3*(((len(node.sequence)-(self.frame-1))//3))))],\n",
    "                                 generic_dna).translate()\n",
    "                        aa_parent = Seq(node.up.sequence[(self.frame-1):(self.frame-1+(3*(((len(node.sequence)-(self.frame-1))//3))))],\n",
    "                                        generic_dna).translate()\n",
    "                        nonsyn = hamming_distance(aa, aa_parent)\n",
    "                        if '*' in aa:\n",
    "                            nstyle['bgcolor'] = 'red'\n",
    "                        if nonsyn > 0:\n",
    "                            nstyle['hz_line_color'] = 'black'\n",
    "                            nstyle['hz_line_width'] = nonsyn\n",
    "                        else:\n",
    "                            nstyle['hz_line_type'] = 1\n",
    "\n",
    "            node.set_style(nstyle)\n",
    "\n",
    "        ts = TreeStyle()\n",
    "        ts.show_leaf_name = False\n",
    "        ts.rotation = 90\n",
    "        def my_layout(node):\n",
    "            if node.frequency > 1:\n",
    "                N = TextFace(node.frequency, fsize=14, fgcolor='black')\n",
    "                N.rotation = -90\n",
    "                faces.add_face_to_node(N, node, 0, position='branch-top')\n",
    "        ts.layout_fn = my_layout\n",
    "        self.tree.ladderize()\n",
    "        self.tree.render(outfile, tree_style=ts)\n",
    "\n",
    "    def write(self, file_name):\n",
    "        '''serialize tree to file'''\n",
    "        with open(file_name, 'wb') as f:\n",
    "            pickle.dump(self, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class CollapsedForest(CollapsedTree):\n",
    "    '''\n",
    "    simply a set of CollapsedTrees, with the same p and q parameters\n",
    "          (4)          (3)\n",
    "         / | \\\\         / \\\\\n",
    "       (3)(1)(2)     (1) (2)\n",
    "           |   \\\\  ,          , ...\n",
    "          (2)  (1)\n",
    "    '''\n",
    "    def __init__(self, params=None, n_trees=None, forest=None):\n",
    "        '''\n",
    "        in addition to p and q, we need number of trees\n",
    "        can also intialize with forest, a list of trees, each an instance of CollapsedTree\n",
    "        '''\n",
    "        CollapsedTree.__init__(self, params=params)\n",
    "        if forest is None and params is None:\n",
    "            raise ValueError('either params or forest (or both) must be provided')\n",
    "        if forest is not None:\n",
    "            if len(forest) == 0:\n",
    "                raise ValueError('passed empty tree list')\n",
    "            if n_trees is not None and len(forest) != n_trees:\n",
    "                raise ValueError('n_trees not consistent with forest')\n",
    "            self.forest = forest\n",
    "        if n_trees is not None:\n",
    "            if type(n_trees) is not int or n_trees < 1:\n",
    "                raise ValueError('number of trees must be at least one')\n",
    "            self.n_trees = n_trees\n",
    "        if n_trees is None and forest is not None:\n",
    "            self.n_trees = len(forest)\n",
    "\n",
    "    def simulate(self):\n",
    "        '''\n",
    "        simulate a forest of collapsed trees given params and number of trees\n",
    "        replaces existing forest data member with simulation result, and returns self\n",
    "        '''\n",
    "        if self.params is None or self.n_trees is None:\n",
    "            raise ValueError('params and n_trees parameters must be defined for simulation')\n",
    "        self.forest = [CollapsedTree(self.params).simulate() for x in range(self.n_trees)]\n",
    "        return self\n",
    "\n",
    "    def l(self, params, sign=1, Vlad_sum=False):\n",
    "        '''\n",
    "        likelihood of params, given forest, and it's gradient wrt params\n",
    "        optional parameter sign must be 1 or -1, with the latter useful for MLE by minimization\n",
    "        if optional parameter Vlad_sum is true, we're doing the Vlad sum for estimating params for\n",
    "        as set of parsimony trees\n",
    "        '''\n",
    "        if self.forest is None:\n",
    "            raise ValueError('forest data must be defined to compute likelihood')\n",
    "        if sign not in (-1, 1):\n",
    "            raise ValueError('sign must be 1 or -1')\n",
    "        # since the l method on the CollapsedTree class returns l and grad_l...\n",
    "        terms = [tree.l(params) for tree in self.forest]\n",
    "        ls = scipy.array([term[0] for term in terms])\n",
    "        grad_ls = scipy.array([term[1] for term in terms])\n",
    "        if Vlad_sum:\n",
    "            # we need to find the smallest derivative component for each\n",
    "            # coordinate, then subtract off to get positive things to logsumexp\n",
    "            grad_l = []\n",
    "            for j in range(len(params)):\n",
    "                i_prime = grad_ls[:,j].argmin()\n",
    "                grad_l.append(grad_ls[i_prime,j] +\n",
    "                              scipy.exp(logsumexp(ls - ls[i_prime],\n",
    "                                                  b=grad_ls[:,j]-grad_ls[i_prime,j]) -\n",
    "                                        logsumexp(ls - ls[i_prime])))\n",
    "            return sign*(-scipy.log(len(ls)) + logsumexp(ls)), sign*scipy.array(grad_l)\n",
    "        else:\n",
    "            return sign*ls.sum(), sign*grad_ls.sum(axis=0)\n",
    "\n",
    "\n",
    "    # NOTE: we get mle() method for free by inheritance/polymorphism magic\n",
    "\n",
    "\n",
    "    def __str__(self):\n",
    "        '''return a string representation for printing'''\n",
    "        return 'params = {}, n_trees = {}\\n'.format(self.params, self.n_trees) + \\\n",
    "                '\\n'.join([str(tree) for tree in self.forest])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def hamming_distance(seq1, seq2):\n",
    "    '''Hamming distance between two sequences of equal length'''\n",
    "    return sum(x != y for x, y in zip(seq1, seq2))\n",
    "\n",
    "\n",
    "def phylip_parse(phylip_outfile, naive=None):\n",
    "    '''parse phylip outfile and return ete trees'''\n",
    "    # parse phylip outfile\n",
    "    outfiledat = [block.split('\\n\\n\\n')[0].split('\\n\\n') for block in open(phylip_outfile, 'r').read().split('From    To     Any Steps?    State at upper node')[1:]]\n",
    "\n",
    "    # ete trees\n",
    "    trees = []\n",
    "    for i, tree in enumerate(outfiledat):\n",
    "        tree_sequence_dict = {}\n",
    "        parent_dict = {}\n",
    "        names = []\n",
    "        for j, block in enumerate(tree):\n",
    "            if j == 0:\n",
    "                for line in block.split('\\n'):\n",
    "                    fields = line.split()\n",
    "                    if len(fields) == 0:\n",
    "                        continue\n",
    "                    name = fields[1]\n",
    "                    names.append(name)\n",
    "                    if fields[0] == 'root':\n",
    "                        seq = ''.join(fields[2:])\n",
    "                        parent = None\n",
    "                    else:\n",
    "                        seq = ''.join(fields[3:])\n",
    "                        parent = fields[0]\n",
    "                    tree_sequence_dict[name] = seq\n",
    "                    parent_dict[name] = parent\n",
    "            else:\n",
    "                for line in block.split('\\n'):\n",
    "                    fields = line.split()\n",
    "                    name = fields[1]\n",
    "                    if fields[0] == 'root':\n",
    "                        seq = ''.join(fields[2:])\n",
    "                    else:\n",
    "                        seq = ''.join(fields[3:])\n",
    "                    tree_sequence_dict[name] += seq\n",
    "\n",
    "        # if integer branch (not weird ambiguous chars)\n",
    "        if set(''.join([tree_sequence_dict[name] for name in names])) == set('ACGT'):\n",
    "            #nodes = dict([(name, Tree(name=(name, tree_sequence_dict[name]), dist=hamming_distance(tree_sequence_dict[name], tree_sequence_dict[parent_dict[name]]) if parent_dict[name] is not None else None)) for name in names])\n",
    "            nodes = {}\n",
    "            for name in names:\n",
    "                node = TreeNode()\n",
    "                node.name = name\n",
    "                node.dist = hamming_distance(tree_sequence_dict[name], tree_sequence_dict[parent_dict[name]]) if parent_dict[name] is not None else 0\n",
    "                node.add_feature('sequence', tree_sequence_dict[node.name])\n",
    "                if node.name == naive:\n",
    "                    node.add_feature('frequency', 0)\n",
    "                elif '_' in node.name:\n",
    "                    node.add_feature('frequency', int(node.name.split('_')[-1]))\n",
    "                    node.name = '_'.join(node.name.split('_')[:-1])\n",
    "                else:\n",
    "                    node.add_feature('frequency', 0)\n",
    "                nodes[name] = node\n",
    "            tree = nodes[names[0]] # naive is first\n",
    "            for name in parent_dict:\n",
    "                if parent_dict[name] is not None:\n",
    "                    nodes[parent_dict[name]].add_child(nodes[name])\n",
    "            # reroot on naive\n",
    "            if naive is not None:\n",
    "                assert len(nodes[naive].children) == 0\n",
    "                assert nodes[naive] in tree.children\n",
    "                tree.remove_child(nodes[naive])\n",
    "                nodes[naive].add_child(tree)\n",
    "                tree.dist = nodes[naive].dist\n",
    "                tree = nodes[naive]\n",
    "                tree.dist = 0\n",
    "\n",
    "            # assert branch lengths make sense\n",
    "            for node in tree.iter_descendants():\n",
    "                assert node.dist == hamming_distance(node.sequence, node.up.sequence)\n",
    "\n",
    "            trees.append(tree)\n",
    "\n",
    "    return trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class MutationModel():\n",
    "    '''a class for a mutation model, and functions to mutate sequences'''\n",
    "    def __init__(self, mutability_file=None, substitution_file=None):\n",
    "        '''initialized with input files of the S5F format'''\n",
    "        if mutability_file is not None and substitution_file is not None:\n",
    "            self.context_model = {}\n",
    "            with open(mutability_file, 'r') as f:\n",
    "                # eat header\n",
    "                f.readline()\n",
    "                for line in f:\n",
    "                    motif, score = line.replace('\"', '').split()[:2]\n",
    "                    self.context_model[motif] = float(score)\n",
    "\n",
    "            # kmer k\n",
    "            self.k = None\n",
    "            with open(substitution_file, 'r') as f:\n",
    "                # eat header\n",
    "                f.readline()\n",
    "                for line in f:\n",
    "                    fields = line.replace('\"', '').split()\n",
    "                    motif = fields[0]\n",
    "                    if self.k is None:\n",
    "                        self.k = len(motif)\n",
    "                        assert self.k % 2 == 1\n",
    "                    else:\n",
    "                        assert len(motif) == self.k\n",
    "                    self.context_model[motif] = (self.context_model[motif], {b:float(x) for b, x in zip('ACGT', fields[1:5])})\n",
    "        else:\n",
    "            self.context_model = None\n",
    "\n",
    "\n",
    "    def mutabilities(self, sequence):\n",
    "        '''returns the mutability of a sequence at each site, along with nucleotide biases'''\n",
    "        assert all(n in 'ACGT' for n in sequence)\n",
    "        sequence_length = len(sequence)\n",
    "        if self.context_model is not None:\n",
    "            # mutabilities of each nucleotide\n",
    "            mutabilities = []\n",
    "            assert sequence_length >= 5\n",
    "            # ambiguous left end motifs\n",
    "            for i in range(self.k//2 + 1):\n",
    "                kmer_suffix = sequence[:(i+self.k//2+1)]\n",
    "                matches = [value for key, value in self.context_model.iteritems() if key.endswith(kmer_suffix)]\n",
    "                len_matches = len(matches)\n",
    "                assert len_matches == 4**(self.k - len(kmer_suffix))\n",
    "                # use mean over matches\n",
    "                mutability = sum(match[0] for match in matches)/len_matches\n",
    "                substitution = {n:sum(d[1][n] for d in matches)/len_matches for n in 'ACGT'}\n",
    "                mutabilities.append((mutability, substitution))\n",
    "            # unambiguous internal kmers\n",
    "            for i in range(self.k//2, sequence_length - self.k//2):\n",
    "                mutabilities.append(self.context_model[sequence[(i-self.k//2):(i+self.k//2+1)]])\n",
    "            # ambiguous right end motifs\n",
    "            for i in range(sequence_length - self.k//2 + 1, sequence_length):\n",
    "                kmer_prefix = sequence[(i-self.k//2):]\n",
    "                matches = [value for key, value in self.context_model.iteritems() if key.startswith(kmer_prefix)]\n",
    "                len_matches = len(matches)\n",
    "                assert len_matches == 4**(self.k - len(kmer_prefix))\n",
    "                # use mean over matches\n",
    "                mutability = sum(match[0] for match in matches)/len_matches\n",
    "                substitution = {n:sum(d[1][n] for d in matches)/len_matches for n in 'ACGT'}\n",
    "                mutabilities.append((mutability, substitution))\n",
    "\n",
    "            return mutabilities\n",
    "        else:\n",
    "            return [(1, dict((n2, 1/3) if n2 is not n else (n2, 0.) for n2 in 'ACGT')) for n in sequence]\n",
    "\n",
    "    def mutate(self, sequence, lambda0=1, frame=1):\n",
    "        '''mutate a sequence, with lamdba0 the baseline mutability'''\n",
    "        sequence_length = len(sequence)\n",
    "        codon_start = frame-1\n",
    "        codon_end = codon_start + 3*((sequence_length - codon_start)//3)\n",
    "        if '*' in Seq(sequence[codon_start:codon_end], generic_dna).translate():\n",
    "            raise RuntimeError('sequence contains stop codon!')\n",
    "\n",
    "        mutabilities = self.mutabilities(sequence)\n",
    "        sequence_mutability = sum(mutability[0] for mutability in mutabilities)/sequence_length\n",
    "        # baseline Poisson\n",
    "        # poisson rate for this sequence (given its relative mutability)\n",
    "        lambda_sequence = sequence_mutability*lambda0\n",
    "        # number of mutations\n",
    "        trials = 10\n",
    "        for trial in range(1, trials+1):\n",
    "            m = scipy.random.poisson(lambda_sequence)\n",
    "            if m <= sequence_length:\n",
    "                break\n",
    "            if trial == trials:\n",
    "                raise RuntimeError('mutations saturating, consider reducing lambda0')\n",
    "        # multinomial sample, with no repeats\n",
    "        p = scipy.array([mutability[0] for mutability in mutabilities])\n",
    "        p = p/p.sum()\n",
    "        mutations = scipy.random.choice(sequence_length, size=m, p=p, replace=False)\n",
    "        # mutute the sites with mutations\n",
    "        # if contains stop codon, try again, up to 10 times\n",
    "        sequence_list = list(sequence) # mutable\n",
    "        for trial in range(1, trials+1):\n",
    "            for i in mutations:\n",
    "                p = [mutabilities[i][1][n] for n in 'ACGT']\n",
    "                assert 0 <= abs(sum(p) - 1.) < 1e-10\n",
    "                sequence_list[i] = 'ACGT'[scipy.random.choice(4, p=p)]\n",
    "            sequence = ''.join(sequence_list)\n",
    "            if '*' not in Seq(sequence[codon_start:codon_end], generic_dna).translate():\n",
    "                break\n",
    "            if trial == trials:\n",
    "                raise RuntimeError('stop codon in simulated sequence on 10 consecustive attempts')\n",
    "        return sequence\n",
    "\n",
    "\n",
    "    def simulate(self, sequence, p=.4, lambda0=1, r=1.):\n",
    "        '''simulate neutral binary branching process with mutation model'''\n",
    "        if p >= .5:\n",
    "            raw_input('WARNING: p = {} is not subcritical, tree termination not garanteed! [ENTER] to proceed'.format(p))\n",
    "        tree = TreeNode()\n",
    "        tree.dist = 0\n",
    "        tree.add_feature('sequence', sequence)\n",
    "        tree.add_feature('terminated', False)\n",
    "        tree.add_feature('frequency', 0)\n",
    "\n",
    "        nodes_unterminated = 1\n",
    "        while nodes_unterminated > 0:\n",
    "            for leaf in tree.iter_leaves():\n",
    "                if not leaf.terminated:\n",
    "                    if scipy.random.random() < p:\n",
    "                        for child_count in range(2):\n",
    "                            mutated_sequence = self.mutate(leaf.sequence, lambda0=lambda0)\n",
    "                            child = TreeNode()\n",
    "                            child.dist = sum(x!=y for x,y in zip(mutated_sequence, leaf.sequence))\n",
    "                            child.add_feature('sequence', mutated_sequence)\n",
    "                            child.add_feature('frequency', 0)\n",
    "                            leaf.add_child(child)\n",
    "                            child.add_feature('terminated' ,False)\n",
    "                        nodes_unterminated += 1\n",
    "                    else:\n",
    "                        leaf.terminated = True\n",
    "                        nodes_unterminated -= 1\n",
    "\n",
    "        # each leaf gets an observation frequency of 1\n",
    "        for node in tree.iter_leaves():\n",
    "            if scipy.random.random() < r:\n",
    "                node.frequency = 1\n",
    "\n",
    "        # return the fine (uncollapsed) tree\n",
    "        return tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def test(args):\n",
    "    '''\n",
    "    test subprogram\n",
    "    checks likelihood against a by-hand calculation for a simple tree, simulates a forest, computes MLE parameters, and plots some sanity check figures to plot_file\n",
    "    command line arguments are p, q, number of trees to simulate, and plot file name\n",
    "    '''\n",
    "    p = args.p\n",
    "    q = args.q\n",
    "    n = args.n\n",
    "    plot_file = args.outbase\n",
    "\n",
    "    if plot_file[-4:] != '.pdf':\n",
    "        plot_file += '.pdf'\n",
    "\n",
    "    print('Let''s check our likelihood against a by-hand calculation for the following simple tree')\n",
    "    # ete tree\n",
    "    parent = TreeNode(format=1)\n",
    "    parent.add_feature('frequency', 2)\n",
    "    parent.name = parent.frequency\n",
    "    child = TreeNode()\n",
    "    child.add_feature('frequency', 1)\n",
    "    child.dist = 1\n",
    "    child.name = child.frequency\n",
    "    parent.add_child(child)\n",
    "    tree = CollapsedTree(tree=parent)\n",
    "    f = 6*p**2*(1-p)**3*q*(1-q)**3\n",
    "    dfdp = 6*(1 - p)**2*p*(-2 + 5*p)*(-1 + q)**3*q #6*q*(1-q)**3*(2*p*(1-p)**3-3*p**2*(1-p)**2)\n",
    "    dfdq = 6*(-1 + p)**3*p**2*(1 - q)**2*(-1 + 4*q) #6*p**2*(1-p)**3*((1-q)**3-3*q*(1-q)**2)\n",
    "    print( '    T =', tree.tree.get_ascii(show_internal=True))\n",
    "    print( '    Summing the probabilities of the two possible fine structures, we have')\n",
    "    print( '    logP =', scipy.log(f))\n",
    "    print(u'    \\u2207logP = ', (dfdp/f, dfdq/f))\n",
    "    print( '    Now, our dynamic programming algorithm gives')\n",
    "    print(u'    logP , \\u2207logP =', tree.l((p, q)))\n",
    "    print('')\n",
    "\n",
    "    print('Simulating a forest of {} trees'.format(n))\n",
    "    forest = CollapsedForest((p, q), n)\n",
    "    print('    true parameters: p = {}, q = {}'.format(p, q))\n",
    "    forest.simulate()\n",
    "\n",
    "    # total leaf counts\n",
    "    total_data = sorted([sum(node.frequency for node in tree.tree.traverse()) for tree in forest.forest])\n",
    "    max_total = max(total_data)\n",
    "    len_total = len(total_data)\n",
    "\n",
    "    totals = []\n",
    "    freq = []\n",
    "    log_prob = []\n",
    "    for x in range(1, max_total+1):\n",
    "        totals.append(x)\n",
    "        freq.append(total_data.count(x))\n",
    "        tmp_tree = TreeNode(format=1)\n",
    "        tmp_tree.add_feature('frequency', x)\n",
    "        log_prob.append(CollapsedTree(tree=tmp_tree).l((p, 0))[0])\n",
    "    theoretical_cdf = scipy.cumsum(scipy.exp(log_prob))\n",
    "    empirical_cdf = scipy.cumsum(freq)/len_total\n",
    "\n",
    "    fig = plt.figure()\n",
    "    fig.set_tight_layout(True)\n",
    "    plt.rc('text', usetex=True)\n",
    "\n",
    "    # plot the empirical and theoretical distribution of total leaf counts\n",
    "\n",
    "    ax = fig.add_subplot(2,2,1)\n",
    "    ax.plot(totals, scipy.exp(log_prob), 'ko', markerfacecolor='none', alpha=.5, label='theoretical PMF')\n",
    "    ax.plot(totals, scipy.array(freq)/len_total, 'k.', label='empirical PMF')\n",
    "    ax.legend(numpoints=1, loc=1, fontsize='small')\n",
    "    ax.set_xlabel('total leaves')\n",
    "    ax.set_ylabel('$\\Pr($total leaves$)$')\n",
    "    ax.set_ylim([0, 1.1])\n",
    "    #ax.set_xscale('log')\n",
    "    #ax.set_yscale('symlog')\n",
    "\n",
    "# uncomment this if you want the CDF\n",
    "#    ax = fig.add_subplot(2,2,2)\n",
    "#    ax.plot(totals, theoretical_cdf, 'ko', markerfacecolor='none', alpha=.5, label='theoretical CDF')\n",
    "#    ax.plot(totals, empirical_cdf, 'k.', label='empirical CDF')\n",
    "#    ax.legend(numpoints=1, loc=4, fontsize='small')\n",
    "#    ax.set_xlabel('number of leaves')\n",
    "#    ax.set_ylim([0, 1.1])\n",
    "\n",
    "\n",
    "    empirical_quantiles = []\n",
    "    theoretical_quantiles = []\n",
    "    for x in total_data:\n",
    "        empirical_quantiles.append(sum(y <= x for y in total_data)/len_total)\n",
    "        to_add = 0.\n",
    "        for y in range(1, x+1):\n",
    "            tmp_tree = TreeNode(format=1)\n",
    "            tmp_tree.add_feature('frequency', y)\n",
    "            to_add += scipy.exp(CollapsedTree(tree=tmp_tree).l((p, 0))[0])\n",
    "        theoretical_quantiles.append(to_add)\n",
    "\n",
    "    ax = fig.add_subplot(2,2,2)\n",
    "    ax.plot(theoretical_quantiles, empirical_quantiles, 'ko', alpha=.1)\n",
    "    ax.plot([0, 1], [0, 1], 'k')\n",
    "    ax.set_title('total leaves')\n",
    "    ax.set_xlabel('theoretical quantiles')\n",
    "    ax.set_ylabel('empirical quantiles')\n",
    "    ax.set_xlim([0, 1])\n",
    "    ax.set_ylim([0, 1])\n",
    "    ax.set_aspect('equal')\n",
    "\n",
    "    mle = forest.mle()\n",
    "\n",
    "    #for tree in forest.forest:\n",
    "    #    print(tree)\n",
    "    print('    MLE parameters:  p = {}, q = {}'.format(*mle.x.tolist()))\n",
    "\n",
    "    # plot the 2-norm of the difference between the gradient and its finite difference approximation\n",
    "    print('computing plot data...')\n",
    "    X, Y = scipy.mgrid[slice(.05, 1, .05),\n",
    "                       slice(.05, 1, .05)]\n",
    "    Z = scipy.zeros((X.shape[0], X.shape[1]))\n",
    "    for i in range(Z.shape[0]):\n",
    "        for j in range(Z.shape[1]):\n",
    "            Z[i, j] = check_grad(lambda x: forest.l(x)[0], lambda x: forest.l(x)[1], (X[i, j], Y[i, j]))\n",
    "\n",
    "    print('done')\n",
    "    ax = fig.add_subplot(2,2,3)\n",
    "    ax.set_title(r'$||\\nabla \\ell(p, q) - \\Delta \\ell(p, q)||_2$')\n",
    "    im = ax.contourf(X, Y, Z, locator=ticker.LogLocator(), cmap='Greys')\n",
    "    ax.set_xlabel(r'$p$')\n",
    "    ax.set_ylabel(r'$q$')\n",
    "    ax.set_aspect('equal')\n",
    "    fig.colorbar(im, ax=ax)\n",
    "\n",
    "\n",
    "    # plot likelihood surface, with true and MLE parameters shown\n",
    "    X, Y = scipy.mgrid[slice(.02, 1, .02),\n",
    "                       slice(.02, 1, .02)]\n",
    "    Z = scipy.zeros((X.shape[0], X.shape[1]))\n",
    "    for i in range(Z.shape[0]):\n",
    "        for j in range(Z.shape[1]):\n",
    "            l, grad_l = forest.l((X[i, j], Y[i, j]))\n",
    "            z = l\n",
    "            Z[i, j] = z\n",
    "    ax = fig.add_subplot(2,2,4)\n",
    "    ax.set_title(r'$\\ell(p, q)$')\n",
    "    contour = ax.contour(X, Y, Z, colors='k', label='likelihood contours')\n",
    "    for c in contour.collections:\n",
    "        c.set_linestyle('solid')\n",
    "\n",
    "    ax.clabel(contour, fontsize=9, inline=1)\n",
    "    ax.plot([p], [q], 'k+', label='true parameters')\n",
    "    ax.plot(mle.x[0], mle.x[1], 'ko', markerfacecolor='none', label='MLE parameters')\n",
    "    ax.set_xlabel(r'$p$')\n",
    "    ax.set_ylabel(r'$q$')\n",
    "    ax.set_aspect('equal')\n",
    "    ax.legend(numpoints = 1, fontsize='small')\n",
    "\n",
    "    plt.savefig(plot_file)\n",
    "    print('plot saved to', plot_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def infer(args):\n",
    "    '''inference subprogram'''\n",
    "    parsimony_forest = CollapsedForest(forest=[CollapsedTree(tree=tree, frame=args.frame) for tree in phylip_parse(args.phylipfile, args.naive)])\n",
    "\n",
    "    if parsimony_forest.n_trees == 1:\n",
    "        warnings.warn('only one parsimony tree reported from dnapars')\n",
    "\n",
    "    print('number of trees with integer branch lengths:', parsimony_forest.n_trees)\n",
    "\n",
    "    # check for unifurcations at root\n",
    "    unifurcations = sum(tree.tree.frequency == 0 and len(tree.tree.children) == 1 for tree in parsimony_forest.forest)\n",
    "    if unifurcations:\n",
    "        print('WARNING: {} trees exhibit unifurcation from root, which is not possible under current model. Such nodes will be ommitted from likelihood calculation'.format(unifurcations))\n",
    "\n",
    "    # fit p and q using all trees\n",
    "    # if we get floating point errors, try a few more times (starting params are random)\n",
    "    max_tries = 1\n",
    "    for tries in range(max_tries):\n",
    "        try:\n",
    "            parsimony_forest.mle(Vlad_sum=True)\n",
    "            break\n",
    "        except FloatingPointError as e:\n",
    "            if tries + 1 < max_tries:\n",
    "                print('floating point error in MLE: {}. Attempt {} of {}. Rerunning with new random start.'.format(e, tries+1, max_tries))\n",
    "            else:\n",
    "                raise\n",
    "        else:\n",
    "            raise\n",
    "\n",
    "    print('params = {}'.format(parsimony_forest.params))\n",
    "\n",
    "    # get likelihoods and sort by them\n",
    "    ls = [tree.l(parsimony_forest.params)[0] for tree in parsimony_forest.forest]\n",
    "    ls, parsimony_forest.forest = zip(*sorted(zip(ls, parsimony_forest.forest), reverse=True))\n",
    "\n",
    "    with open(args.outbase+'.inference.parsimony_forest.p', 'wb') as f:\n",
    "        pickle.dump(parsimony_forest, f)\n",
    "\n",
    "    print('tree\\talleles\\tlogLikelihood')\n",
    "    for i, (l, collapsed_tree) in enumerate(zip(ls, parsimony_forest.forest), 1):\n",
    "        alleles = len(collapsed_tree.tree)\n",
    "        print('{}\\t{}\\t{}'.format(i, alleles, l))\n",
    "        collapsed_tree.render(args.outbase+'.inference.{}.svg'.format(i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def simulate(args):\n",
    "    '''simulation subprogram'''\n",
    "    if args.lambda0 is None:\n",
    "        args.lambda0 = max([1, int(.01*len(args.sequence))])\n",
    "    args.sequence = args.sequence.upper()\n",
    "    mutation_model = MutationModel(args.mutability, args.substitution)\n",
    "    size = 0\n",
    "    while size < args.n:\n",
    "        # this loop makes us resimulate if we got backmutations\n",
    "        trial = 1\n",
    "        while trial < 10:\n",
    "            try:\n",
    "                tree = mutation_model.simulate(args.sequence, p=args.p, lambda0=args.lambda0, r=args.r)\n",
    "                collapsed_tree = CollapsedTree(tree=tree, frame=args.frame) # <-- this will fail if backmutations\n",
    "                break\n",
    "            except RuntimeError:\n",
    "                trial += 1\n",
    "                continue\n",
    "            else:\n",
    "                raise\n",
    "        if trial == 10:\n",
    "            raise RuntimeError('repeated sequences in collapsed tree on {} attempts'.format(trial))\n",
    "        size = sum(node.frequency for node in tree)\n",
    "    with open(args.outbase+'.simulation.fasta', 'w') as f:\n",
    "        f.write('> naive\\n')\n",
    "        f.write(args.sequence+'\\n')\n",
    "        i = 0\n",
    "        for leaf in tree.iter_leaves():\n",
    "            if leaf.frequency != 0:# and '*' not in Seq(leaf.sequence, generic_dna).translate():\n",
    "                i += 1\n",
    "                f.write('> seq{}\\n'.format(i))\n",
    "                f.write(leaf.sequence+'\\n')\n",
    "                leaf.name = 'seq{}'.format(i)\n",
    "    print(i, 'simulated observed sequences')\n",
    "    collapsed_tree.write( args.outbase+'.simulation.collapsed_tree.p')\n",
    "    collapsed_tree.render(args.outbase+'.simulation.collapsed_tree.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def validate(args):\n",
    "    with open(args.truetree, 'rb') as f:\n",
    "        true_tree = pickle.load(f)\n",
    "    with open(args.parfor, 'rb') as f:\n",
    "        parsimony_forest = pickle.load(f)\n",
    "\n",
    "    # NOTE: the unrooted_trees flag is needed because, for some reason, the RF\n",
    "    #       function sometimes thinks the collapsed trees are unrooted and barfs\n",
    "    distances, likelihoods = zip(*[(true_tree.tree.robinson_foulds(tree.tree, attr_t1='sequence', attr_t2='sequence', unrooted_trees=True)[0],\n",
    "                                    tree.l(parsimony_forest.params)[0]) for tree in parsimony_forest.forest])\n",
    "\n",
    "    df = pd.DataFrame({'distance':distances, 'log-likelihood':likelihoods})\n",
    "    df.to_csv(args.outbase+'.validation.tsv', sep='\\t', index=False)\n",
    "    sns.regplot(x='log-likelihood', y='distance', data=df).get_figure().savefig(args.outbase+'.validation.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def main():\n",
    "    import argparse\n",
    "\n",
    "    parser = argparse.ArgumentParser(description='germinal center tree inference and simulation')\n",
    "    subparsers = parser.add_subparsers(help='which program to run')\n",
    "\n",
    "    # parser for test subprogram\n",
    "    parser_test = subparsers.add_parser('test',\n",
    "                                        formatter_class=argparse.ArgumentDefaultsHelpFormatter,\n",
    "                                        help='run tests on library functions')\n",
    "    parser_test.add_argument('--p', type=float, default=.4, help='branching probability for test mode')\n",
    "    parser_test.add_argument('--q', type=float, default=.5, help='mutation probability for test mode')\n",
    "    parser_test.add_argument('--n', type=int, default=1, help='forest size for test mode')\n",
    "    parser_test.set_defaults(func=test)\n",
    "\n",
    "    # parser for inference subprogram\n",
    "    parser_infer = subparsers.add_parser('infer',\n",
    "                                         formatter_class=argparse.ArgumentDefaultsHelpFormatter,\n",
    "                                         help='likelihood ranking of parsimony trees')\n",
    "    parser_infer.add_argument('--naive', type=str, default=None, help='name of naive sequence (outgroup root)')\n",
    "    parser_infer.add_argument('phylipfile', type=str, help='dnapars outfile (verbose output with sequences at each site)')\n",
    "    parser_infer.set_defaults(func=infer)\n",
    "\n",
    "    # parser for simulation subprogram\n",
    "    parser_sim = subparsers.add_parser('simulate',\n",
    "                                       formatter_class=argparse.ArgumentDefaultsHelpFormatter,\n",
    "                                       help='neutral model gctree simulation')\n",
    "    parser_sim.add_argument('sequence', type=str, help='seed naive nucleotide sequence')\n",
    "    parser_sim.add_argument('mutability', type=str, help='path to mutability model file')\n",
    "    parser_sim.add_argument('substitution', type=str, help='path to substitution model file')\n",
    "    parser_sim.add_argument('--p', type=float, default=.49, help='branching probability')\n",
    "    parser_sim.add_argument('--lambda0', type=float, default=None, help='baseline mutation rate')\n",
    "    parser_sim.add_argument('--r', type=float, default=1., help='sampling probability')\n",
    "    parser_sim.add_argument('--n', type=int, default=1, help='minimum simulation size')\n",
    "    parser_sim.set_defaults(func=simulate)\n",
    "\n",
    "    # parser for validation subprogram\n",
    "    parser_val = subparsers.add_parser('validate',\n",
    "                                       formatter_class=argparse.ArgumentDefaultsHelpFormatter,\n",
    "                                       help='validate results of inference on simulation data')\n",
    "    parser_val.add_argument('truetree', type=str, help='.p file containing true tree')\n",
    "    parser_val.add_argument('parfor', type=str, help='.p file containing parsimony forest from inference')\n",
    "    parser_val.set_defaults(func=validate)\n",
    "\n",
    "    # a common outbase parameter\n",
    "    for subparser in [parser_test, parser_infer, parser_sim, parser_val]:\n",
    "        subparser.add_argument('--outbase', type=str, default='gctree.out', help='output file base name')\n",
    "\n",
    "    # a common parameter for the inference and simulation subprograms\n",
    "    for subparser in [parser_infer, parser_sim]:\n",
    "        subparser.add_argument('--frame', type=int, default=None, help='codon frame')\n",
    "\n",
    "    args = parser.parse_args()\n",
    "    args.func(args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17 simulated observed sequences\n"
     ]
    }
   ],
   "source": [
    "# Simulate the command line input:\n",
    "import sys\n",
    "# command_string = 'simulate -h'\n",
    "command_string = 'simulate \\\n",
    "GGCGCAGGACTGTTGAAGCCTTCGGAGACCCTGTCCCTCACCTGCGCTGTCTATGGTGGGTCCTTCAGTGGTTACTACTGGAGCTGGATCCGCCAGCCCCCAGGGAAGGGGCTGG\\\n",
    "AGTGGATTGGGGAAATCAATCATAGTGGAAGCACCAACTACAACCCGTCCCTCAAGAGTCGAGTCACCATATCAGTAGACACGTCCAAGAACCAGTTCTCCCTGAAGCTGAGCTC\\\n",
    "TGTGACCGCCGCGGACACGGCTGTGTATTACTGTGCGAGAGAGGGAGTGGGAGCTACTACGGCTTACTACTACGGTATGGACGTCTGGGGCCAAGGGACCACGGTCACCGTC \\\n",
    "S5F/Mutability.csv S5F/Substitution.csv --n 10'\n",
    "sys.argv = command_string.split()\n",
    "sys.argv.insert(0, 'gctree.py')\n",
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gctree.py\n",
      "simulate\n",
      "GGCGCAGGACTGTTGAAGCCTTCGGAGACCCTGTCCCTCACCTGCGCTGTCTATGGTGGGTCCTTCAGTGGTTACTACTGGAGCTGGATCCGCCAGCCCCCAGGGAAGGGGCTGGAGTGGATTGGGGAAATCAATCATAGTGGAAGCACCAACTACAACCCGTCCCTCAAGAGTCGAGTCACCATATCAGTAGACACGTCCAAGAACCAGTTCTCCCTGAAGCTGAGCTCTGTGACCGCCGCGGACACGGCTGTGTATTACTGTGCGAGAGAGGGAGTGGGAGCTACTACGGCTTACTACTACGGTATGGACGTCTGGGGCCAAGGGACCACGGTCACCGTC\n",
      "S5F/Mutability.csv\n",
      "S5F/Substitution.csv\n",
      "--n\n",
      "10\n"
     ]
    }
   ],
   "source": [
    "print('\\n'.join(sys.argv))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
